src/rng/backends/curand/mrg32k3a.cpp

```diff
@@ -0,0 +1,956 @@
+                        reinterpret_cast<float*>(ih.get_native_mem<cl::sycl::backend::cuda>(acc));
+                    curandStatus_t status;
+                    CURAND_CALL(curandGenerateUniform, status, engine_, r_ptr, n);
+                });
+            })
+            .wait_and_throw();
+        range_transform_fp<float>(queue_, distr.a(), distr.b(), n, r);
+    }
+
+    virtual void generate(
+        const oneapi::mkl::rng::uniform<double, oneapi::mkl::rng::uniform_method::standard>& distr,
+        std::int64_t n, cl::sycl::buffer<double, 1>& r) override {
+        queue_
+            .submit([&](cl::sycl::handler& cgh) {
+                auto acc = r.get_access<cl::sycl::access::mode::read_write>(cgh);
+                cgh.codeplay_host_task([=](cl::sycl::interop_handle ih) {
+                    auto r_ptr =
+                        reinterpret_cast<double*>(ih.get_native_mem<cl::sycl::backend::cuda>(acc));
+                    curandStatus_t status;
+                    CURAND_CALL(curandGenerateUniformDouble, status, engine_, r_ptr, n);
+                });
+            })
+            .wait_and_throw();
+        range_transform_fp<double>(queue_, distr.a(), distr.b(), n, r);
+    }
+
+    virtual void generate(const oneapi::mkl::rng::uniform<
+                              std::int32_t, oneapi::mkl::rng::uniform_method::standard>& distr,
+                          std::int64_t n, cl::sycl::buffer<std::int32_t, 1>& r) override {
+        cl::sycl::buffer<std::uint32_t, 1> ib(n);
+        queue_
+            .submit([&](cl::sycl::handler& cgh) {
+                auto acc = ib.get_access<cl::sycl::access::mode::read_write>(cgh);
+                cgh.codeplay_host_task([=](cl::sycl::interop_handle ih) {
+                    auto ib_ptr = reinterpret_cast<std::uint32_t*>(
+                        ih.get_native_mem<cl::sycl::backend::cuda>(acc));
+                    curandStatus_t status;
+                    CURAND_CALL(curandGenerate, status, engine_, ib_ptr, n);
+                });
+            })
+            .wait_and_throw();
+        range_transform_int<std::int32_t>(queue_, distr.a(), distr.b(), n, ib, r);
+    }
+
+    virtual void generate(
+        const oneapi::mkl::rng::uniform<float, oneapi::mkl::rng::uniform_method::accurate>& distr,
+        std::int64_t n, cl::sycl::buffer<float, 1>& r) override {
+        queue_
+            .submit([&](cl::sycl::handler& cgh) {
+                auto acc = r.get_access<cl::sycl::access::mode::read_write>(cgh);
+                cgh.codeplay_host_task([=](cl::sycl::interop_handle ih) {
+                    auto r_ptr =
+                        reinterpret_cast<float*>(ih.get_native_mem<cl::sycl::backend::cuda>(acc));
+                    curandStatus_t status;
+                    CURAND_CALL(curandGenerateUniform, status, engine_, r_ptr, n);
+                });
+            })
+            .wait_and_throw();
+        range_transform_fp<float>(queue_, distr.a(), distr.b(), n, r);
```

Comment:
For oneMKL RNG we have an additional boundary check in the accurate method: does it makes sense to add range_transfor_fp_accurate, which would perform such check?
described in the spec [modes of random number generation](https://spec.oneapi.com/versions/latest/elements/oneMKL/source/domains/rng/distributions.html)

Reply:
Thanks for the link! However, I don't really understand the difference between "accurate" and "fast"; it is not clear from the description:

"_...For example, random numbers obtained from the generator of continuous distribution that is uniform on interval [a,b] belong to this interval irrespective of what a and b values may be. Fast mode provides high performance generation and also guarantees that generated random numbers belong to the definitional domain except for some specific values of distribution parameters._"

The ad hoc `range_transform_X` functions implemented adjust the generated pseudorandom numbers are within the specified range. Since cuRAND doesn't have this functionality built-in, and the unit tests specify a custom range for the numbers, the transformations are necessary here.

The unclear piece is "_...except for some specific values of distribution parameters_". What criteria are these?

Reply:
For an accurate method, there are no criteria for any specific values of distribution parameters, we just have an additional check, like for uniform:
```
if(res < a)
   res = a;
 if(res > b)
    res = b;
```
for any distribution's parameters